Use `math/rand/v2` in `k8s.io/apimachinery/pkg/util/rand`

tklauser opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind cleanup

#### What this PR does / why we need it:

The `math/rand/v2` Go standard libary package was added in Go 1.22[^1][^2][^3]. The new package provides a way to construct PRNG sources which can be accessed concurrently, making the locking in `k8s.io/apimachinery/pkg/util/rand` unnecessary.

[^1]: https://go.dev/blog/randv2
[^2]: https://go.dev/doc/go1.22#math_rand_v2
[^3]: https://pkg.go.dev/math/rand/v2

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
N/A

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/rand/rand.go:
```diff
@@ -18,63 +18,44 @@ limitations under the License.
-	defer rng.Unlock()
-
-	rng.rand = rand.New(rand.NewSource(seed))
+	rng = rand.New(rand.NewPCG(uint64(seed), 0))
```
previously, this was threadsafe, but now concurrent calls to set a package variable could cause a data race with other calls to Seed or any of the functions that read / call methods on `rng`

you could resolve this by making rng an atomic.Pointer that readers can Load() and this function can Store(), or by making a package RWMutex readers can RLock/RUnlock and this function can Lock/Unlock

tklauser replied:
Fixed by making `rng` an `atomic.Pointer`. Thank you for the review!

tklauser committed:
Use math/rand/v2 in k8s.io/apimachinery/pkg/util/rand

The math/rand/v2 Go standard libary package was added in Go
1.22[^1][^2][^3]. The new package provides a way to construct PRNG
sources which can be accessed concurrently, making the locking in
k8s.io/apimachinery/pkg/util/rand unnecessary.

[^1]: https://go.dev/blog/randv2
[^2]: https://go.dev/doc/go1.22#math_rand_v2
[^3]: https://pkg.go.dev/math/rand/v2
